testutils/gke/request.go

```diff
@@ -0,0 +1,126 @@
+	GKEVersion string
+
+	// ClusterName: name of the cluster
+	ClusterName string
+
+	// MinNodes: the minimum number of nodes of the cluster
+	MinNodes int64
+
+	// MaxNodes: the maximum number of nodes of the cluster
+	MaxNodes int64
+
+	// NodeType: node type of the cluster, e.g. n1-standard-4, n1-standard-8
+	NodeType string
+
+	// Region: region of the cluster, e.g. us-west1, us-central1
+	Region string
+
+	// Zone: default is none, must be provided together with region
+	Zone string
+
+	// Addons: cluster addons to be added to cluster, such as istio
+	Addons []string
+}
+
+// DeepCopy will make a deepcopy of the request struct.
+func (r *Request) DeepCopy() *Request {
+	return &Request{
+		Project:     r.Project,
+		GKEVersion:  r.GKEVersion,
+		ClusterName: r.ClusterName,
+		MinNodes:    r.MinNodes,
+		MaxNodes:    r.MaxNodes,
+		NodeType:    r.NodeType,
+		Region:      r.Region,
+		Zone:        r.Zone,
+		Addons:      r.Addons,
+	}
+}
+
+// NewCreateClusterRequest returns a new CreateClusterRequest that can be used in gcloud SDK.
+func NewCreateClusterRequest(request *Request) (*container.CreateClusterRequest, error) {
+	if request.Project == "" {
+		return nil, errors.New("project name cannot be empty")
+	}
+	if request.ClusterName == "" {
+		return nil, errors.New("cluster name cannot be empty")
+	}
+	if request.MinNodes <= 0 {
+		return nil, errors.New("min nodes must be larger than 1")
+	}
+	if request.MinNodes > request.MaxNodes {
+		return nil, errors.New("min nodes cannot be larger than max nodes")
+	}
+	if request.NodeType == "" {
+		return nil, errors.New("node type cannot be empty")
+	}
+
+	if request.GKEVersion == "" {
+		request.GKEVersion = defaultGKEVersion
+	}
+
+	return &container.CreateClusterRequest{
+		Cluster: &container.Cluster{
+			NodePools: []*container.NodePool{
+				{
+					Name:             "default-pool",
+					InitialNodeCount: request.MinNodes,
+					Autoscaling: &container.NodePoolAutoscaling{
+						Enabled:      true,
+						MinNodeCount: request.MinNodes,
+						MaxNodeCount: request.MaxNodes,
+					},
+					Config: &container.NodeConfig{
+						MachineType: request.NodeType,
+					},
+				},
+			},
+			Name: request.ClusterName,
+			// The default cluster version is not latest, has to explicitly
+			// set it as "latest"
+			InitialClusterVersion: request.GKEVersion,
+			// Installing addons after cluster creation takes at least 5
+			// minutes, so install addons as part of cluster creation, which
+			// doesn't seem to add much time on top of cluster creation
+			AddonsConfig: GetAddonsConfig(request.Addons),
+			// Equivalent to --enable-basic-auth, so that user:pass can be
+			// later on retrieved for setting up cluster roles. Use the
+			// default username from gcloud command, the password will be
+			// automatically generated by GKE SDK
+			MasterAuth: &container.MasterAuth{Username: "admin"},
+		},
+		ProjectId: request.Project,
```

Comment:
I wonder if we have to set `ProjectId` here, since we will also pass the project name in the `CreateCluster` function.

Reply:
I'm not sure, You can try it out, there are a bunch of fields going to be deprecated, this could be one of the them

Reply:
It's deprecated, I removed it and `CreateCluster` still works.